File: src/contracts/provenance.ts
/**
 * Provenance — the second of the two things git does NOT give us.
 * git has commit author/message, but not "where did this knowledge come from,
 * what produced it". This is the minimal envelope collected by GET_PROVENANCE.
 */

export type OriginKind =
  | "SOURCE"
  | "OBSERVATION"
  | "EVIDENCE"
  | "ASSERTION"
  | "DEFINITION"
  | "DERIVATION";

export interface ProvenanceEnvelope {
  readonly originKind: OriginKind;
  /** Who. */
  readonly actorRef?: string;
  /** What activity. */
  readonly activityRef?: string;
  /** Sources. */
  readonly sourceRefs?: readonly string[];
  /** Supporting evidence. */
  readonly evidenceRefs?: readonly string[];
  /** Required for DERIVATION only. */
  readonly inputViewReadVersionRef?: string;
  /** Required for DERIVATION only. */
  readonly algorithm?: {
    readonly derivationSpecRef?: string;
    readonly modelRef?: string;
    readonly codeHash?: string;
  };
  readonly producedAt?: string;
}

File: kernel/provenance.go
package kernel

type OriginKind string

const (
	OriginSource      OriginKind = "SOURCE"
	OriginObservation OriginKind = "OBSERVATION"
	OriginEvidence    OriginKind = "EVIDENCE"
	OriginAssertion   OriginKind = "ASSERTION"
	OriginDefinition  OriginKind = "DEFINITION"
	OriginDerivation  OriginKind = "DERIVATION"
)

type AlgorithmRef struct {
	DerivationSpecRef string `json:"derivationSpecRef,omitempty"`
	ModelRef          string `json:"modelRef,omitempty"`
	CodeHash          string `json:"codeHash,omitempty"`
}

func (a AlgorithmRef) Identified() bool {
	return a.DerivationSpecRef != "" || a.ModelRef != "" || a.CodeHash != ""
}

type ProvenanceEnvelope struct {
	OriginKind              OriginKind    `json:"originKind"`
	ActorRef                string        `json:"actorRef,omitempty"`
	ActivityRef             string        `json:"activityRef,omitempty"`
	SourceRefs              []string      `json:"sourceRefs,omitempty"`
	EvidenceRefs            []string      `json:"evidenceRefs,omitempty"`
	InputViewReadVersionRef string        `json:"inputViewReadVersionRef,omitempty"`
	Algorithm               *AlgorithmRef `json:"algorithm,omitempty"`
	ProducedAt              string        `json:"producedAt,omitempty"`
}

func ValidateProvenance(p *ProvenanceEnvelope) error {
	if p == nil || p.OriginKind != OriginDerivation {
		return nil
	}
	identified := p.Algorithm != nil && p.Algorithm.Identified()
	if p.InputViewReadVersionRef == "" || !identified {
		return Fail(ErrPreconditionFailed, "DERIVATION provenance requires a fixed input ViewReadVersion and algorithm identity")
	}
	return nil
}
